### ISSUE TITLE
[lang] Expose thread index in @ti.kernel

### ISSUE THREAD

**Issue by yolo2themoon** [taichi-dev/taichi#2725]
**Concisely describe the proposed feature**
Getting the thread index is a common operation in parallel computing, and exposing the thread index in `@ti.kernel` with something like `ti.block_idx()` and `ti.thread_idx()` might be useful for some applications.

**Describe the solution you'd like (if any)**
For the CUDA backend, all I can think of is that the thread index and block index are stored in special registers and can be accessed directly. As for the frontend, not sure if we can refer to `ti.block_dim(block_dim)` to implement `ti.thread_idx()`

**Comment by lucywsq** [comment]
Hi @Luke-Skycrawler  please join this slack channel, then we would respond to your problems on time. We can communicate faster as well.

**Comment by yolo2themoon** [comment]
#3319  has been merged into the master. Many thanks to @Luke-Skycrawler . cc @lucywsq
